pkg/sysinfo/hugepages.go

```diff
@@ -0,0 +1,157 @@
+
+const (
+	SysDevicesNode = "/sys/devices/system/node"
+)
+
+const (
+	HugepageSize2Mi = 2048
+	HugepageSize1Gi = 1048576
+)
+
+type Handle struct {
+	Root string
+}
+
+func (hnd Handle) SysDevicesNodes() string {
+	return filepath.Join(hnd.Root, SysDevicesNode)
+}
+
+func (hnd Handle) SysDevicesNodesNodeNth(nodeID int) string {
+	return filepath.Join(hnd.Root, SysDevicesNode, fmt.Sprintf("node%d", nodeID))
+}
+
+// TODO review
+type Hugepages struct {
+	NodeID int
+	SizeKB int
+	Total  int
+}
+
+type PerNUMACounters map[int]int64
+
+func GetHugepageCounters(hnd Handle, getHPs func(Handle) ([]*Hugepages, error)) (map[string]PerNUMACounters, error) {
+	numaCounters := make(map[string]PerNUMACounters)
+	hugepages, err := getHPs(hnd)
+	if err != nil {
+		return numaCounters, err
+	}
+
+	for _, hpage := range hugepages {
+		resourceName := hpResourceNameFromSize(hpage.SizeKB)
+		numaDevs, ok := numaCounters[resourceName]
+		if !ok {
+			numaDevs = make(PerNUMACounters)
+		}
+
+		numaDevs[hpage.NodeID] = int64(hpage.Total)
+		numaCounters[resourceName] = numaDevs
+	}
+
+	return numaCounters, nil
+}
+
+func GetHugepages(hnd Handle) ([]*Hugepages, error) {
+	entries, err := ioutil.ReadDir(hnd.SysDevicesNodes())
+	if err != nil {
+		return nil, err
+	}
+
+	hugepages := []*Hugepages{}
+	for _, entry := range entries {
+		entryName := entry.Name()
+		if entry.IsDir() && strings.HasPrefix(entryName, "node") {
+			nodeID, err := strconv.Atoi(entryName[4:])
+			if err != nil {
+				klog.Warningf("cannot detect the node ID for %q", entryName)
+				continue
+			}
+			nodeHugepages, err := HugepagesForNode(hnd, nodeID)
+			if err != nil {
+				klog.Warningf("cannot find the hugepages on NUMA node %d: %v", nodeID, err)
+				continue
+			}
+			hugepages = append(hugepages, nodeHugepages...)
+		}
+	}
+	return hugepages, nil
+}
+
+func HugepagesForNode(hnd Handle, nodeID int) ([]*Hugepages, error) {
+	path := filepath.Join(
+		hnd.SysDevicesNodesNodeNth(nodeID),
+		"hugepages",
+	)
+	hugepages := []*Hugepages{}
+
+	entries, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+	for _, entry := range entries {
+		entryName := entry.Name()
+		entryPath := filepath.Join(path, entryName)
+		var hugepageSizeKB int
+		if n, err := fmt.Sscanf(entryName, "hugepages-%dkB", &hugepageSizeKB); n != 1 || err != nil {
+			klog.Warningf("malformed hugepages entry %q", entryName)
+			continue
+		}
+
+		totalCount, err := readIntFromFile(filepath.Join(entryPath, "nr_hugepages"))
+		if err != nil {
+			klog.Warningf("cannot read from %q: %v", entryPath, err)
+			continue
+		}
+
+		hugepages = append(hugepages, &Hugepages{
+			NodeID: nodeID,
+			SizeKB: hugepageSizeKB,
+			Total:  totalCount,
+		})
+	}
+
+	return hugepages, nil
+}
+
+func readIntFromFile(path string) (int, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return -1, err
+	}
+	return strconv.Atoi(strings.TrimSpace(string(data)))
+}
+
+func hpResourceNameFromSize(sizeKB int) string {
```

Comment:
I wonder if this is the best way, why not convert the symbolic name to the size instead? You could then just traverse the /sys dir and expose everything you find in there without caring about what sizes are supported.

Reply:
done